### PR TITLE
Implement Double Backward for `cdist` and `pdist`

### PR DESCRIPTION
**Pull request by colalb1**
Adds second-order backward for `torch.cdist` and `F.pdist`, which previously only supported first-order gradients. All `create_graph=True` calls use raised "the derivative for '_cdist_backward' is not implemented".

The enabling change renames the first argument of `_cdist_backward`/`_pdist_backward` from `grad` to `grad_output`, since the autograd codegen reserves grad for the incoming gradient. This flows through `native_functions.yaml`, the meta registrations, the generated AOTI c-shims, and a BC allowlist entry. `derivatives.yaml` then wires each backward to a new analytic double backward in `FunctionsManual`. `_cdist_backward_backward` specializes on $p=(0, 1, 2, inf, general)$ with negative powers masked at exact zeros. `_pdist_backward_backward` symmetrizes the packed inputs and reuses the `cdist` double backward. `supports_gradgrad` is enabled for both OpInfos, with $p<1$ covered by dedicated tests that compare the analytic result against an autograd reference.

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -305,6 +305,179 @@ std::tuple<Tensor, Tensor> _euclidean_dist_backward(
...
+
+  auto idx = at::triu_indices(n, n, 1, self.options().dtype(at::kLong));
+  auto row = idx.select(0, 0);
+  auto col = idx.select(0, 1);
+  auto lin_ab = row * n + col;
+  auto lin_ba = col * n + row;
+  
+  auto scatter_sym = [&](const Tensor& packed) {
+    auto full = at::zeros({n * n}, packed.options());
+    full.put_(lin_ab, packed);
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp]: Hmm, we might need to fix a vmap bug here first? Or at least add more xfails

Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp]: Ah ok, I will look into this shortly.

Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp]: @Skylion007 I've addressed this. Please let me know if there are other changes to be made.

Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp]: The issue was that `full.put_(...)` was trying to write a batched tensor `packed.options()` to an unbatched tensor `full`.

The new solution using `index_put` and `index` returns a new tensor instead of changing the argument, so the batching rule can allocate new space.

The old and new methods are numerically identical, so this should reduce the test failures.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -305,6 +305,174 @@ std::tuple<Tensor, Tensor> _euclidean_dist_backward(
...
+    const Tensor& pdist,
+    std::array<bool, 3> output_mask) {
+  const bool need_go = output_mask[0];
+  const bool need_self = output_mask[1];
+  const bool need_pdist = output_mask[2];
+  if (!grad.defined() || !(need_go || need_self || need_pdist)) {
+    return std::make_tuple(Tensor(), Tensor(), Tensor());
+  }
+
+  int64_t n = self.size(0);
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp:433]: Could be an auto sym_size couldn't it? The scatter_sym doesn't look right?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 3]
```diff
@@ -305,6 +305,174 @@ std::tuple<Tensor, Tensor> _euclidean_dist_backward(
...
+        std::move(grad_grad_output),
+        std::move(grad_self),
+        std::move(grad_pdist));
+  }
+
+  auto idx = at::triu_indices(n, n, 1, self.options().dtype(at::kLong));
+  torch::List<std::optional<Tensor>> ij({idx.select(0, 0), idx.select(0, 1)});
+
+  auto scatter_sym = [&](const Tensor& packed) {
+    auto upper = at::zeros({n, n}, packed.options()).index_put(ij, packed);
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp]: Isn't there a zeros_sym?

File: test/test_autograd.py [hunk 4]
```diff
@@ -14083,6 +14083,102 @@ def forward(self, x):
         self.assertEqual(model.a.grad.device, torch.device("cpu"))
         self.assertEqual(model.b.grad.device, torch.device("cpu"))
 
+    @skipIfTorchDynamo("compiled autograd + gradgradcheck undefined-grad mode")
```
Comment (Skylion007) [inline test/test_autograd.py]: Wait, pretty sure this is a fixable bug though?

Comment (colalb1) [reply test/test_autograd.py]: Addressed

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 3 above)
Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp]: Changed from `at::zeros({n, n}, packed.options())` to `new_zeros({n, n})`.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 2 above)
Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp:433]: Changed the name to `to_symmetric` for better clarity since the function means symmetric instead of symbolic.

Comment (Skylion007) [reply torch/csrc/autograd/FunctionsManual.cpp:433]: We have a PR landing soon that should preventing need to do this for SymmMem, might still be needed for DTensor or NestedTensor though?

File: tools/autograd/derivatives.yaml [hunk 5]
```diff
@@ -1360,11 +1358,8 @@
   x1: _cdist_backward(grad.contiguous(), x1, x2, p, result)
   x2: _cdist_backward(grad.mT().contiguous(), x2, x1, p, result.mT().contiguous())
 
-- name: _cdist_backward(Tensor grad, Tensor x1, Tensor x2, float p, Tensor cdist) -> Tensor
-  grad: not_implemented("_cdist_backward")
-  x1: not_implemented("_cdist_backward")
-  x2: not_implemented("_cdist_backward")
-  cdist: not_implemented("_cdist_backward")
+- name: _cdist_backward(Tensor grad_output, Tensor x1, Tensor x2, float p, Tensor cdist) -> Tensor
```
Comment (Skylion007) [inline tools/autograd/derivatives.yaml:1361]: What's the point of changing the signature from grad to grad_output? Why not make it consistent with else like ggrad etc?

Comment (colalb1) [reply tools/autograd/derivatives.yaml:1361]: @Skylion007

In `derivative.yaml`, my understanding was that the `grad` name is reserved for the gradient being returned from a function, and implementing the gradient of `_cdist_backward()` requires both the `grad` (second-order gradient, output of `_cdist_backward_backward()`) and `grad_output` (first-order gradient, output of `_cdist_backward()` and input of `_cdist_backward_backward()`).

In hindsight, it would have been better to use `ggrad` as the name for the returned gradient.

Comment (colalb1) [reply tools/autograd/derivatives.yaml:1361]: `grad` collides with the codegen's reserved name for the upstream gradient (`const auto& grad = grads[0];`), so an input named `grad` makes any implemented derivative fail to compile. That collision is why this entry was not_implemented. `grad_output` is the established convention for `*_backward` first args, so this makes the signature more consistent

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 2 above)
Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp:433]: Here `n` is only called at `triu_indices(n, n, ...)` and `new_zeros({n, n})`. `triu_indices` takes a concrete int, so `sym_size(0)` would just get `guard_int`'d since it can't stay symbolic through this path. I can switch to `sym_size(0)` if you want it for DTensor dispatch, but I don't believe it will give real dynamic-shape support until `triu_indices` is symbolic. Do you see this as blocking for this PR, or a follow-up once your PR lands?

File: tools/autograd/derivatives.yaml (hunk 5 above)
Comment (colalb1) [reply tools/autograd/derivatives.yaml:1361]: @Skylion007 Would it be fine to leave it as `grad_output` given this logic?

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 2 above)
Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp:433]: @Skylion007 Could you link the blocking PR?

I could implement it if need be, just lmk. Thanks.

Comment (Skylion007) [reply torch/csrc/autograd/FunctionsManual.cpp:433]: Sigh went back into draft state: https://github.com/pytorch/pytorch/pull/188650

**Review by Skylion007 — APPROVED** [review]

File: test/test_autograd.py [hunk 6]
```diff
@@ -14083,6 +14083,113 @@ def forward(self, x):
         self.assertEqual(model.a.grad.device, torch.device("cpu"))
         self.assertEqual(model.b.grad.device, torch.device("cpu"))
 
+    @dtypes(torch.double)
```
Comment (Skylion007) [inline test/test_autograd.py:14086]: This doesn't cause issues on MPS?

Comment (colalb1) [reply test/test_autograd.py:14086]: @Skylion007 It did not when I tested locally + in the CI. I assume it works given the test passed on both.

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -23637,7 +23637,7 @@ def sample_inputs_abs(op_info, device, dtype, requires_grad, op_kwargs=None, **k
         sample_inputs_func=sample_inputs_pdist,
         dtypes=floating_types(),
         supports_out=False,
-        supports_gradgrad=False,
+        supports_gradgrad=True,
         skips=(
```
[the remark this answers is not in the document]
Comment (colalb1) [reply torch/testing/_internal/common_methods_invocations.py:23635]: addressed

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -13795,7 +13795,7 @@ def sample_inputs_abs(op_info, device, dtype, requires_grad, op_kwargs=None, **k
            dtypes=floating_types(),
            dtypesIfMPS=floating_types_and(torch.float16, torch.bfloat16),
            supports_out=False,
-           supports_gradgrad=False,
+           supports_gradgrad=True,
            assert_autodiffed=False,
            sample_inputs_func=sample_inputs_cdist,
            skips=(
```
[the remark this answers is not in the document]
Comment (colalb1) [reply torch/testing/_internal/common_methods_invocations.py:13801]: addressed

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 2 above)
Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp:433]: @Skylion007 Ah ok.

If there is anything else that should change in this PR s.t. [188650](https://github.com/pytorch/pytorch/pull/188650) is the only blocking constraint, please let me know. Otherwise I will wait for 188650 to merge.

Thanks

Comment (ngimel) [reply torch/csrc/autograd/FunctionsManual.cpp:433]: This PR has nothing to do with symmetric memory, `to_symmetric` produces symmetric matrices
